keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md

```diff
@@ -0,0 +1,478 @@
+
+The implementation consists of two parts:
+
+1. **cadvisor**: Add `FreePages *uint64` field to `HugePagesInfo` struct, populated
+   from sysfs. Uses pointer with `omitempty` to distinguish between "0 free" and
+   "data unavailable".
+
+2. **kubelet Memory Manager**: Add `verifyOSHugepagesAvailability()` function
+   called during `Allocate()` that compares requested hugepages against OS-reported
+   free hugepages from cadvisor's machine info.
+
+### cadvisor Changes
+
+```go
+type HugePagesInfo struct {
+    // huge page size (in kB)
+    PageSize uint64 `json:"page_size"`
+    // number of huge pages
+    NumPages uint64 `json:"num_pages"`
+    // number of free huge pages (nil if unavailable)
+    FreePages *uint64 `json:"free_pages,omitempty"`
+}
+```
+
+The `FreePages` field is populated by reading from:
+```
+/sys/devices/system/node/node<N>/hugepages/hugepages-<size>kB/free_hugepages
+```
+
+### Memory Manager Changes
+
+During `Allocate()` in the Static policy:
+
+```go
+func (p *staticPolicy) verifyOSHugepagesAvailability(
+    machineState state.NUMANodeMap,
+    pod *v1.Pod,
+    container *v1.Container,
+) error {
+    // For each hugepage size requested by the container:
+    // 1. Get the OS-reported free hugepages from cadvisor machine info
+    // 2. Compare against the requested amount
+    // 3. Return error if insufficient
+}
+```
+
+The verification:
+- Only runs when the Static policy is enabled
+- Only checks hugepage resources (not regular memory)
+- Aggregates free hugepages across candidate NUMA nodes
+- Returns admission error if insufficient free hugepages
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+- Existing Memory Manager unit tests cover allocation logic
+- cadvisor tests cover sysfs reading functionality
+
+##### Unit tests
+
+- `pkg/kubelet/cm/memorymanager`: Add tests for `verifyOSHugepagesAvailability()`
+  - Test successful verification when free hugepages >= requested
+  - Test rejection when free hugepages < requested
+  - Test graceful handling when FreePages is nil (sysfs unavailable)
+  - Test per-NUMA node verification
+
+##### Integration tests
+
+- Test Memory Manager with mocked cadvisor returning various FreePages values
+- Test admission flow with hugepage verification enabled/disabled
+
+##### e2e tests
+
+- Test pod admission when hugepages are available
+- Test pod rejection when hugepages are exhausted
+- Test that rejected pods can be admitted after hugepages are freed
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind `MemoryManagerHugepagesVerification` feature gate
+- Unit tests for verification logic
+- Documentation for feature gate and behavior
+
+#### Beta
+
+- E2e tests demonstrating correct behavior
+- Metrics for verification failures
```

Comment:
I'd strongly prefer to have these at least partially available at alpha stage

Reply:
Agreed. I'll update the alpha graduation criteria to include at least partial e2e test coverage for the core scenario.